### PR TITLE
feat(snowflake): improve snowflake lineage perf and memory, push down to snowflake

### PR DESCRIPTION
**Pull request by mayurinehate**
## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py [hunk 1]
```diff
@@ -292,111 +292,45 @@ def operational_data_for_time_window(
...
-            PARTITION BY view_name,
-            downstream_table_name {", downstream_table_columns" if include_column_lineage else ""}
-            ORDER BY
-              query_start_time DESC
-          ) = 1
+        GROUP BY
+            concat(
+                referencing_database, '.', referencing_schema,
+                '.', referencing_object_name
+            )
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:333]: I quickly tried this and using the earlier defined alias "DOWNSTREAM_TABLE_NAME" works here, will update this.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py [hunk 2]
```diff
@@ -598,3 +532,172 @@ def usage_per_object_per_time_bucket_for_time_window(
...
+                "('Table', 'External table', 'View', 'Materialized view')"
+            )
+        else:
+            allowed_upstream_table_domains = "('Table', 'External table')"
+        return f"""
+        WITH column_lineage_history AS (
+            SELECT
+                r.value : "objectName" :: varchar AS upstream_table_name,
+                r.value : "objectDomain" :: varchar AS upstream_table_domain,
+                -- r.value:"columns" AS upstream_table_columns,
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:553]: yes, will remove

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py [hunk 3]
```diff
@@ -598,3 +532,172 @@ def usage_per_object_per_time_bucket_for_time_window(
...
+                lateral flatten(input => wcols.value : "directSourceColumns", outer => true) wcols_directSources
+            WHERE
+                r.value : "objectId" IS NOT NULL
+                AND w.value : "objectId" IS NOT NULL
+                AND w.value : "objectName" NOT LIKE '%.GE_TMP_%'
+                AND w.value : "objectName" NOT LIKE '%.GE_TEMP_%'
+                AND t.query_start_time >= to_timestamp_ltz({start_time_millis}, 3)
+                AND t.query_start_time < to_timestamp_ltz({end_time_millis}, 3)
+                AND upstream_table_domain in {allowed_upstream_table_domains}
+                AND downstream_table_domain = 'Table'
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:583]: yes

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py [hunk 4]
```diff
@@ -598,3 +532,172 @@ def usage_per_object_per_time_bucket_for_time_window(
...
+            ),
+        table_upstreams AS (
+            SELECT
+                downstream_table_name,
+                ANY_VALUE(downstream_table_domain) as downstream_table_domain,
+                ARRAY_UNIQUE_AGG(
+                OBJECT_CONSTRUCT(
+                    'upstream_object_name', upstream_table_name,
+                    'upstream_object_domain', upstream_table_domain
+                )
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:593]: yes

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py [hunk 5]
```diff
@@ -598,3 +532,172 @@ def usage_per_object_per_time_bucket_for_time_window(
...
+            SELECT
+                downstream_table_name,
+                downstream_column_name,
+                query_start_time,
+                ARRAY_AGG(
+                OBJECT_CONSTRUCT(
+                    'object_name', upstream_column_table_name,
+                    'object_domain', upstream_column_object_domain,
+                    'column_name', upstream_column_name
+                    )
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:610]: yes

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py [hunk 6]
```diff
@@ -598,3 +532,172 @@ def usage_per_object_per_time_bucket_for_time_window(
...
+                ) as upstream_columns_for_job
+            FROM
+                column_lineage_history
+            WHERE
+                upstream_column_name is not null
+                and upstream_column_table_name is not null
+            GROUP BY
+                downstream_table_name,
+                downstream_column_name,
+                query_start_time -- or query_id
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:620]: makes sense. will do this.

File: metadata-ingestion/tests/integration/snowflake/snowflake_golden.json [hunk 7]
```diff
@@ -3587,6 +3587,30 @@
         "runId": "snowflake-2022_06_07-17_00_00"
     }
 },
+{
+    "entityType": "dataset",
+    "entityUrn": "urn:li:dataset:(urn:li:dataPlatform:snowflake,test_db.test_schema.view_2,PROD)",
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/tests/integration/snowflake/snowflake_golden.json:3592]: right. Earlier golden test setup was incorrectly setting view_2 as upstream of table_2 in view_dependencies query, which was supposed to give upstream dependencies of view. Corrected that one, hence this change.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py [hunk 8]
```diff
@@ -598,3 +532,172 @@ def usage_per_object_per_time_bucket_for_time_window(
...
+            SELECT
+                downstream_table_name,
+                downstream_column_name,
+                ARRAY_UNIQUE_AGG(upstream_columns_for_job) as upstreams
+            FROM
+                column_upstream_jobs
+            GROUP BY
+                downstream_table_name,
+                downstream_column_name
+            )
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:623]: It appears that window function and group by can not be used in same SQL query, so this will not work.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_lineage.py [hunk 9]
```diff
@@ -407,72 +331,112 @@ def _process_external_lineage_result_row(self, db_row):
...
-                upstream_table_name,
-                db_row["UPSTREAM_TABLE_COLUMNS"],
-                db_row["DOWNSTREAM_TABLE_COLUMNS"],
-            ),
-            self.config.include_column_lineage,
-        )
-        self.report.num_table_to_table_edges_scanned += 1
-        logger.debug(f"Lineage[Table(Down)={key}]:Table(Up)={self._lineage_map[key]}")
+            upstreams.append(
+                UpstreamClass(
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_lineage.py:377]: It it is okay, I would prefer to do it in follow up PR than here.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py [hunk 10]
```diff
@@ -598,3 +532,172 @@ def usage_per_object_per_time_bucket_for_time_window(
...
+        SELECT
+            table_upstreams.downstream_table_name AS "DOWNSTREAM_TABLE_NAME",
+            ANY_VALUE(
+                table_upstreams.downstream_table_domain
+            ) AS "DOWNSTREAM_TABLE_DOMAIN",
+            ANY_VALUE(
+                table_upstreams.upstream_tables
+            ) AS "UPSTREAM_TABLES",
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:631]: will do.

[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py]: done

**Comment by mayurinehate** [comment]
> I know we want to keep functions short and reduce duplicate code but I found this pretty hard to follow. I think separating the query parsing with the actual workunit generation may help in reducing this complexity.

I understand what you are saying. I have already separated query parsing (`get_upstreams_from_query_result_row`) and actual workunit generation (`_create_upstream_lineage_workunit`). I am not entirely sure how to make this simpler, given the complexity of fine grained lineage itself. It it is okay, let's do this in follow up PR. Also it would be great if you can help with that one.

**Comment by mayurinehate** [comment]
> If we're confident on the correctness due to connector tests and confident on the performance (not sure how we test this? -- directly relies on snowflake because snowflake does the majority of the processing here) then I think this is good enough to ship.

I am quite confident about correctness of these changes as I have already validated old upstreams against new upstreams via connector tests. As you mentioned , performance is entirely dependent on snowflake, however I also believe, pushing the aggregations to snowflake (which is optimised to to such stuff) is far better (in terms of performance) than doing these inside datahub ingestion process.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py [hunk 11]
```diff
@@ -133,7 +138,7 @@ def get_all_tags_in_database_without_propagation(db_name: str) -> str:
         domain as "DOMAIN"
         FROM snowflake.account_usage.tag_references
         WHERE (object_database = '{db_name}' OR object_name = '{db_name}')
-        AND domain in ('DATABASE', 'SCHEMA', 'TABLE', 'COLUMN')
+        AND lower(domain) in {allowed_object_domains}
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:141]: I'm a little worried that this will be significantly slower. e.g. if `domain` has an index, we won't be able to use the index right? Because we're applying a transformation on the column. I understand wanting to reuse `SnowflakeObjectDomain` but I'd rather not potentially compromise performance just to get rid of magic strings

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py [hunk 12]
```diff
@@ -120,6 +122,9 @@ def get_all_tags_on_object_with_propagation(
 
     @staticmethod
     def get_all_tags_in_database_without_propagation(db_name: str) -> str:
+
+        allowed_object_domains = f"('{SnowflakeObjectDomain.DATABASE}', '{SnowflakeObjectDomain.SCHEMA}', '{SnowflakeObjectDomain.TABLE}', '{SnowflakeObjectDomain.COLUMN}')"
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:126]: Can you make sure that this produces the appropriate string? I thought when I tested this out myself on some bigquery changes, the `str` of a `StrEnum` value like `SnowflakeObjectDomain` didn't produce what I wanted:

```
In [1]: from datahub.ingestion.source.snowflake.constants import SnowflakeObjectDomain

In [2]: str(SnowflakeObjectDomain.DATABASE)
Out[2]: 'SnowflakeObjectDomain.DATABASE'
```

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py [hunk 13]
```diff
@@ -389,6 +391,9 @@ def usage_per_object_per_time_bucket_for_time_window(
         objects_column = (
             "BASE_OBJECTS_ACCESSED" if use_base_objects else "DIRECT_OBJECTS_ACCESSED"
         )
+
+        allowed_object_domains = f"('{SnowflakeObjectDomain.TABLE}', '{SnowflakeObjectDomain.EXTERNAL_TABLE}', '{SnowflakeObjectDomain.VIEW}', '{SnowflakeObjectDomain.MATERIALIZED_VIEW}')"
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:395]: It seems like we reuse the following sets of domains:
- Table, External table, View, Materialized view
- Table, External table

What do you think about putting them in variables, e.g.:
```
table_and_view_domains_filter = f"('{SnowflakeObjectDomain.TABLE}', '{SnowflakeObjectDomain.EXTERNAL_TABLE}', '{SnowflakeObjectDomain.VIEW}', '{SnowflakeObjectDomain.MATERIALIZED_VIEW}')"
table_domains_filter = f"('{SnowflakeObjectDomain.TABLE}', '{SnowflakeObjectDomain.EXTERNAL_TABLE}')"
```

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (asikowitz) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py]: This looks great, thanks!

**Review by asikowitz — COMMENTED** [review]
This looks really good, thanks Mayuri! I think the changes made in response to my magic string comment are not going to work, because `StrEnum` don't behave as expected, and I'm worried about performance of doing `lower` checks. I'm not familiar with Snowflake's information schema views or query performance, so will defer to you if you think otherwise, but personally I'd rather keep the magic strings or just make new constants rather than do `lower` comparisons in queries. Otherwise, as you are confident on correctness, I think this is good to go. When we merge, we should not only run connector tests but make a pre-release and test this on a larger dataset, to measure performance + ensure the same workunits are produced by the old and new implementations. Do you think it's at all worth it to keep the old implementation in case something is not handled correctly by this one, e.g. behind a feature flag?

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py (hunk 11 above)
Comment (asikowitz) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:141]: Perhaps we can use `SnowflakeObjectDomain.____.value.capitalize()` instead?

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py (hunk 12 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:126]: Using these in formatted string returns correct results

```
>>> from datahub.ingestion.source.snowflake.constants import SnowflakeObjectDomain
>>> f"{SnowflakeObjectDomain.DATABASE}"
'database'

```

Comment (asikowitz) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:126]: Wow, today I learned that f strings call __format__ and not __str__. Very surprised that they're different... if I want to format a value to a string is the most effective way really `f"{value}"`?

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py (hunk 13 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:395]: yeah, that would work, making this change now.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py (hunk 11 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:141]: That's a valid concern and highly dependent on how snowflake handles this. We are using [snowflake account usage views ](https://docs.snowflake.com/en/sql-reference/account-usage#account-usage-views) here and snowflake does not have concept of indices but rather uses [clustering and micro-partitions](https://docs.snowflake.com/en/user-guide/tables-micro-partitions) that impact query performance. Let me look around how snowflake handles such case sensitive comparisons and if there is a better way.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py (hunk 12 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:126]: looks like that.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py (hunk 13 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:395]: done

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py (hunk 11 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:141]: I am going to remove lower casing in query , assuming the worst case.

**Comment by mayurinehate** [comment]
> Do you think it's at all worth it to keep the old implementation in case something is not handled correctly by this one, e.g. behind a feature flag?

I've added the flag `use_legacy_lineage_method` which is enabled by default (old behavior by default). We should switch it to False once we've full confidence on new optimised lineage method.

**Review by asikowitz — APPROVED** [review]
LGTM! One question about tests but otherwise looks great

File: metadata-ingestion/tests/integration/snowflake/test_snowflake_failures.py [hunk 14]
```diff
@@ -55,6 +55,7 @@ def snowflake_pipeline_config(tmp_path):
                 schema_pattern=AllowDenyPattern(allow=["test_db.test_schema"]),
                 include_view_lineage=False,
                 include_usage_stats=False,
+                use_legacy_lineage_method=False,
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/tests/integration/snowflake/test_snowflake_failures.py:58]: Added tests. Also did a small refractor  - calling legacy snowflake lineage module as `snowflake_lineage_legacy.py` and new snowflake lineage module as `snowflake_lineage_v2`.

**Review by asikowitz — APPROVED** [review]
Still looks good to me!
